mlflow/sagemaker/__init__.py

```diff
@@ -338,30 +451,146 @@ def _deploy(role, image_url, app_name, model_s3_path, run_id, region_name):
-                'InitialVariantWeight': 1,
-            },
+                'InitialVariantWeight': 1 
+            }
         ],
         Tags=[
             {
                 'Key': 'app_name',
-                'Value': app_name,
+                'Value': endpoint_name,
             },
         ],
     )
-    eprint("endpoint_config_arn: %s" % endpoint_config_response["EndpointConfigArn"])
+    
+    eprint("Created endpoint configuration with arn: %s" % endpoint_config_response["EndpointConfigArn"])
     endpoint_response = sage_client.create_endpoint(
-        EndpointName=app_name,
+        EndpointName=endpoint_name,
         EndpointConfigName=config_name,
         Tags=[],
     )
-    eprint("endpoint_arn: %s" % endpoint_response["EndpointArn"])
+    eprint("Created endpoint with arn: %s" % endpoint_response["EndpointArn"])
+
+def _update_sagemaker_endpoint(endpoint_name, image_url, model_s3_path, run_id, mode, archive, role, sage_client, s3_client):
+    """
+    :param image_url: URL of the ECR-hosted docker image the model is being deployed into
+    :param model_s3_path: s3 path where we stored the model artifacts
+    :param run_id: RunId that generated this model
+    :param mode: either `add` or `replace`.
+    :param archive: If True, any pre-existing SageMaker application resources that become inactive
+                    (i.e. as a result of deploying in `replace` mode) will be preserved. If False,
+                    these resources will be deleted.
+    :param role: SageMaker execution ARN role
+    :param sage_client: A boto3 client for SageMaker
+    :param s3_client: A boto3 client for S3 
+    """
+    assert mode in [DEPLOYMENT_MODE_ADD, DEPLOYMENT_MODE_REPLACE], \
+            "Unrecognized mode: `{md}` for deployment to pre-existing application".format(md=mode)
+   
+    endpoint_info = sage_client.describe_endpoint(EndpointName=endpoint_name)
+    endpoint_arn = endpoint_info["EndpointArn"]
+    deployed_config_name = endpoint_info["EndpointConfigName"]
+    deployed_config_info = sage_client.describe_endpoint_config(EndpointConfigName=deployed_config_name)
+    deployed_config_arn = deployed_config_info["EndpointConfigArn"]
+    deployed_production_variants = deployed_config_info["ProductionVariants"]
+
+    eprint("Found active endpoint with arn: {earn}. Updating...".format(earn=endpoint_arn))
+
+    new_model_name = _get_sagemaker_model_name(endpoint_name)
+    new_model_response = sage_client.create_model(
+        ModelName=new_model_name,
+        PrimaryContainer={
+            'ContainerHostname': 'mlflow-serve-%s' % new_model_name,
+            'Image': image_url,
+            'ModelDataUrl': model_s3_path,
+            'Environment': {},
+        },
+        ExecutionRoleArn=role,
+        Tags=[{'Key': 'run_id', 'Value': str(run_id)}, ],
+    )
+    eprint("Created new model with arn: %s" % new_model_response["ModelArn"])
+
+    if mode == DEPLOYMENT_MODE_ADD:
+        new_model_weight = 0
```

Comment:
We might want to note that the newly-added model has a weight of 0 (this means no traffic gets routed to the model right?) in the doc for `deploy` / briefly mention how the user can route traffic to the new model